asf_lifetime_cost_model/pipeline/lifetime_cost_calculator.py

```diff
@@ -0,0 +1,354 @@
+        """
+        if purchase_year >= config.get("installation_year_max"):
+            raise ValueError(f"Purchase year must be before {config.get('installation_year_max')}.")
+
+        if heating_system not in ["ashp", "boiler"]:
+            raise ValueError(
+                f"Unsupported heating system: {heating_system}. Supported heating systems are `ashp` and `boiler`."
+            )
+
+        if (heating_system == "boiler") and (subsidy_model_or_input_values != "no subsidy"):
+            raise ValueError("Subsidy model is only applicable for air source heat pumps. Choose 'no subsidy'.")
+
+
+        if installation_costs is None:
+            installation_costs = (
+                self.ashp_installation_costs.copy() if heating_system == "ashp" else self.boiler_installation_costs.copy()
+            )
+        installation_costs = get_installation_cost(
+            costs_data=installation_costs, heating_system=heating_system, decile=decile
+        )
+
+        # Apply installation cost reduction assumption
+        cost_reduction_data = create_cost_reduction_data(annual_cost_reduction=annual_cost_reduction)
+        cost_reduction_value = cost_reduction_data[purchase_year]
+        installation_costs["installation_cost"] = installation_costs["installation_cost"] * cost_reduction_value
+
+        # ASHP with preset subsidy model
+        if (type(subsidy_model_or_input_values) is str) and (heating_system == "ashp"):
+            subsidy_data = self.ashp_subsidy_options_data.copy()
+            subsidy_value = get_ashp_subsidy_value(
+                subsidy_data=subsidy_data, subsidy_model=subsidy_model_or_input_values, purchase_year=purchase_year
+            )
+            installation_costs["installation_cost"] = installation_costs["installation_cost"] - subsidy_value
+        # ASHP with custom subsidy model
+        elif heating_system == "ashp":  # dictionary provided
+            if purchase_year not in subsidy_model_or_input_values.keys():
+                raise ValueError(f"Please provide subsidy value for purchase year: {purchase_year}")
+            installation_costs["installation_cost"] = (
+                installation_costs["installation_cost"] - subsidy_model_or_input_values[purchase_year]
+            )
+
+        if purchase_with_loan:
+            if life_span is None:
+                raise ValueError("Please provide `life_span` in years when purchasing with loans.")
+
+            if loan_interest_rate < 0.0 or loan_interest_rate > 1.0:
+                raise ValueError("Loan interest rate should be provided as a decimal between 0 and 1.")
+
+            installation_costs["installation_cost"] = installation_costs.apply(
+                lambda x: compute_cost_with_loan(
+                    upfront_cost=x["installation_cost"], life_span=life_span, loan_interest_rate=loan_interest_rate
+                ),
+                axis=1,
+            )
+
```

Comment:
```suggestion
            installation_costs["loan_interest"] = installation_costs.apply(
                lambda x: (
                    compute_cost_with_loan(
                        upfront_cost=x["installation_cost"],
                        life_span=life_span,
                        loan_interest_rate=loan_interest_rate,
                    )
                    - x["installation_cost"]
                ),
                axis=1,
            )

```

I suggest adding the loan interest amount as a separate column to make our life easier when we need to separate out installation cost vs. loan payment in the output figures of the app.